diff --git a/cmd/decorate/decorate.go b/cmd/decorate/decorate.go
index e36c6ea46..1e8ebe0bc 100644
--- a/cmd/decorate/decorate.go
+++ b/cmd/decorate/decorate.go
@@ -7,7 +7,6 @@ import (
 	"fmt"
 	"go/format"
 	"io"
-	"maps"
 	"os"
 	"reflect"
 	"slices"
@@ -17,7 +16,6 @@ import (
 
 	"github.com/Masterminds/sprig/v3"
 	"github.com/evcc-io/evcc/api"
-	combinations "github.com/mxschmitt/golang-combinations"
 	"github.com/samber/lo"
 	"github.com/spf13/pflag"
 	"golang.org/x/tools/imports"
@@ -34,9 +32,13 @@ var srcTmpl string
 //go:embed header.tpl
 var header string
 
+type dependentTypes struct {
+	T, Dependent string
+}
+
 type funcStruct struct {
-	Signature, Function, VarName, ReturnTypes string
-	Params                                    []string
+	Signature, Function, VarName, ReturnTypes, BaseType, ShortType string
+	Params                                                         []string
 }
 
 type typeStruct struct {
@@ -84,80 +86,33 @@ func init() {
 	}
 }
 
-// hasIntersection returns if the slices intersect
-func hasIntersection[T comparable](a, b []T) bool {
-	for _, el := range a {
-		if slices.Contains(b, el) {
-			return true
-		}
-	}
-	return false
-}
-
-func getCombinations(combos []string) [][]string {
-	validCombos := make([][]string, 0)
-	sortedDependents := slices.Sorted(maps.Keys(dependents))
-
-COMBO:
-	for _, c := range combinations.All(combos) {
-		// order the cases for generation
-		for _, master := range sortedDependents {
-			details := dependents[master]
-			// prune combinations where ...
-			// - master is part of the decorators
-			// - master is not part of the currently evaluated combination
-			// - details are part of the currently evaluated combination
-			// ... and remove details from the combination
-			if slices.Contains(combos, master) && !slices.Contains(c, master) && hasIntersection(c, details) {
-				c = lo.Without(c, details...)
-
-				if len(c) == 0 {
-					continue COMBO
-				}
-			}
-		}
-
-		// prune duplicates
-		for _, v := range validCombos {
-			if slices.Equal(v, c) {
-				continue COMBO
-			}
-		}
-
-		validCombos = append(validCombos, c)
-	}
-
-	return validCombos
-}
-
 func getTemplate(dtypes []reflect.Type, types map[string]typeStruct, combos []string) *template.Template {
 	tmpl, err := template.New("gen").Funcs(sprig.FuncMap()).Funcs(template.FuncMap{
-		// contains checks if slice contains string
-		"contains": slices.Contains[[]string, string],
-		// ordered returns a slice of funcStruct ordered by dynamicType
-		"ordered": func() []funcStruct {
-			ordered := make([]funcStruct, 0)
+		"dependents": func() []dependentTypes {
+			dt := make([]dependentTypes, 0)
+			dtypesStrings := lo.Map(dtypes, func(t reflect.Type, _ int) string {
+				return t.String()
+			})
+
 			for _, t := range dtypes {
-				for _, f := range types[getTypeImport(t)].Functions {
-					ordered = append(ordered, f)
+				for key, values := range dependents {
+					if slices.Contains(dtypesStrings, key) && slices.Contains(values, t.String()) {
+						dt = append(dt, dependentTypes{T: types[key].Functions[0].VarName, Dependent: types[t.String()].Functions[0].VarName})
+					}
 				}
 			}
-			return ordered
+
+			return dt
 		},
-		"requiredType": func(c []string, typ string) bool {
-			for master, details := range dependents {
-				// exclude combinations where ...
-				// - master is part of the decorators
-				// - master is not part of the currently evaluated combination
-				// - details are part of the currently evaluated combination
-				if slices.Contains(combos, master) && !slices.Contains(c, master) && slices.Contains(details, typ) {
-					return false
+		// orderedParams returns a slice of funcStruct ordered by dynamicType
+		"orderedParams": func() []funcStruct {
+			orderedParams := make([]funcStruct, 0)
+			for _, t := range dtypes {
+				for _, f := range types[getTypeImport(t)].Functions {
+					orderedParams = append(orderedParams, f)
 				}
 			}
-			return true
-		},
-		"empty": func() []string {
-			return nil
+			return orderedParams
 		},
 	}).Parse(srcTmpl)
 
@@ -214,6 +169,8 @@ func generate(out io.Writer, functionName, baseType string, dtypes []reflect.Typ
 				Function:    m.Name,
 				Params:      params,
 				ReturnTypes: fmt.Sprintf("(%s)", strings.Join(returns, ",")),
+				BaseType:    t.String(),
+				ShortType:   t.Name(),
 			})
 		}
 
@@ -243,12 +200,11 @@ func generate(out io.Writer, functionName, baseType string, dtypes []reflect.Typ
 		Types               map[string]typeStruct
 		Combinations        [][]string
 	}{
-		Function:     functionName,
-		BaseType:     baseType,
-		ShortBase:    shortBase,
-		ReturnType:   returnType,
-		Types:        types,
-		Combinations: getCombinations(combos),
+		Function:   functionName,
+		BaseType:   baseType,
+		ShortBase:  shortBase,
+		ReturnType: returnType,
+		Types:      types,
 	}
 
 	return getTemplate(dtypes, types, combos).Execute(out, vars)
diff --git a/cmd/decorate/decorate.tpl b/cmd/decorate/decorate.tpl
index 87271417e..f198f02bf 100644
--- a/cmd/decorate/decorate.tpl
+++ b/cmd/decorate/decorate.tpl
@@ -1,56 +1,34 @@
-{{define "case"}}
-	{{- $combo := .Combo}}
-	{{- $prefix := .Prefix}}
-	{{- $and := false}}
+func {{.Function}}(base {{.BaseType}}{{range orderedParams}}, {{.VarName}} {{.Signature}}{{end}}) {{.ReturnType}} {
+{{- $prefix := .Function}}
+{{- range dependents}}
+	if {{.T}} == nil {
+		{{.Dependent}} = nil
+	}
+	{{end}}
 
-	{{- range $typ, $def := .Types}}
-		{{- if requiredType $combo $typ}}
-			{{- range $def.Functions}}
-				{{- if $and}} &&{{else}}{{$and = true}}{{end}} {{.VarName}} {{if contains $combo $typ}}!={{else}}=={{end}} nil
-			{{- end}}
-		{{- end}}
-	{{- end}}:
-		return &struct {
-			{{.BaseType}}
-{{- range $typ, $def := .Types}}
-	{{- if contains $combo $typ}}
-			{{$typ}}
-	{{- end}}
-{{- end}}
-		}{
-			{{.ShortBase}}: base,
-{{- range $typ, $def := .Types}}
-	{{- if contains $combo $typ}}
-			{{$def.ShortType}}: &{{$prefix}}{{$def.ShortType}}Impl{
-				{{- range $def.Functions}}
-				{{.VarName}}: {{.VarName}},
-				{{- end}}
-			},
-	{{- end}}
-{{- end}}
-		}
-{{- end -}}
+	caps := make(map[reflect.Type]any)
 
-func {{.Function}}(base {{.BaseType}}{{range ordered}}, {{.VarName}} {{.Signature}}{{end}}) {{.ReturnType}} {
-{{- $basetype := .BaseType}}
-{{- $shortbase := .ShortBase}}
-{{- $prefix := .Function}}
-{{- $types := .Types}}
-{{- $and := false}}
-	switch {
-	case {{- range $typ, $def := .Types}}
-		{{- if requiredType empty $typ}}
-			{{- range $def.Functions}}
-				{{- if $and}} &&{{else}}{{$and = true}}{{end}} {{.VarName}} == nil
-			{{- end}}
-		{{- end}}
-	{{- end}}:
+	{{range orderedParams}}
+	if {{.VarName}} != nil {
+		caps[reflect.TypeFor[{{.BaseType}}]()] = &{{$prefix}}{{.ShortType}}Impl{ {{.VarName}}: {{.VarName}} }	
+	}
+	{{end}}
+
+	if len(caps) == 0 {
 		return base
-{{range $combo := .Combinations}}
-	case {{- template "case" dict "BaseType" $basetype "Prefix" $prefix "ShortBase" $shortbase "Types" $types "Combo" $combo}}
-{{end}}	}
+	}
+
+	return &{{.Function}}Capable{ {{.ShortBase}}: base, caps: caps}
+}
+
+type {{.Function}}Capable struct {
+	{{.BaseType}}
+	caps map[reflect.Type]any
+}
 
-	return nil
+func (d *{{.Function}}Capable) Capability(typ reflect.Type) (any, bool) {
+	c, ok := d.caps[typ]
+	return c, ok
 }
 
 {{range $element := .Types -}}
diff --git a/cmd/decorate/decorate_decorators.go b/cmd/decorate/decorate_decorators.go
index a78e63b5a..b7f0d6abd 100644
--- a/cmd/decorate/decorate_decorators.go
+++ b/cmd/decorate/decorate_decorators.go
@@ -3,87 +3,45 @@ package main
 // Code generated by github.com/evcc-io/evcc/cmd/tools/decorate.go. DO NOT EDIT.
 
 import (
+	"reflect"
+
 	"github.com/evcc-io/evcc/api"
 )
 
 func decorateTest(base api.Charger, meterEnergy func() (float64, error), phaseSwitcher func(int) error, phaseGetter func() (int, error)) api.Charger {
-	switch {
-	case meterEnergy == nil && phaseSwitcher == nil:
-		return base
+	if phaseSwitcher == nil {
+		phaseGetter = nil
+	}
+
+	caps := make(map[reflect.Type]any)
+
+	if meterEnergy != nil {
+		caps[reflect.TypeFor[api.MeterEnergy]()] = &decorateTestMeterEnergyImpl{meterEnergy: meterEnergy}
+	}
+
+	if phaseSwitcher != nil {
+		caps[reflect.TypeFor[api.PhaseSwitcher]()] = &decorateTestPhaseSwitcherImpl{phaseSwitcher: phaseSwitcher}
+	}
 
-	case meterEnergy != nil && phaseSwitcher == nil:
-		return &struct {
-			api.Charger
-			api.MeterEnergy
-		}{
-			Charger: base,
-			MeterEnergy: &decorateTestMeterEnergyImpl{
-				meterEnergy: meterEnergy,
-			},
-		}
-
-	case meterEnergy == nil && phaseGetter == nil && phaseSwitcher != nil:
-		return &struct {
-			api.Charger
-			api.PhaseSwitcher
-		}{
-			Charger: base,
-			PhaseSwitcher: &decorateTestPhaseSwitcherImpl{
-				phaseSwitcher: phaseSwitcher,
-			},
-		}
-
-	case meterEnergy != nil && phaseGetter == nil && phaseSwitcher != nil:
-		return &struct {
-			api.Charger
-			api.MeterEnergy
-			api.PhaseSwitcher
-		}{
-			Charger: base,
-			MeterEnergy: &decorateTestMeterEnergyImpl{
-				meterEnergy: meterEnergy,
-			},
-			PhaseSwitcher: &decorateTestPhaseSwitcherImpl{
-				phaseSwitcher: phaseSwitcher,
-			},
-		}
-
-	case meterEnergy == nil && phaseGetter != nil && phaseSwitcher != nil:
-		return &struct {
-			api.Charger
-			api.PhaseGetter
-			api.PhaseSwitcher
-		}{
-			Charger: base,
-			PhaseGetter: &decorateTestPhaseGetterImpl{
-				phaseGetter: phaseGetter,
-			},
-			PhaseSwitcher: &decorateTestPhaseSwitcherImpl{
-				phaseSwitcher: phaseSwitcher,
-			},
-		}
-
-	case meterEnergy != nil && phaseGetter != nil && phaseSwitcher != nil:
-		return &struct {
-			api.Charger
-			api.MeterEnergy
-			api.PhaseGetter
-			api.PhaseSwitcher
-		}{
-			Charger: base,
-			MeterEnergy: &decorateTestMeterEnergyImpl{
-				meterEnergy: meterEnergy,
-			},
-			PhaseGetter: &decorateTestPhaseGetterImpl{
-				phaseGetter: phaseGetter,
-			},
-			PhaseSwitcher: &decorateTestPhaseSwitcherImpl{
-				phaseSwitcher: phaseSwitcher,
-			},
-		}
+	if phaseGetter != nil {
+		caps[reflect.TypeFor[api.PhaseGetter]()] = &decorateTestPhaseGetterImpl{phaseGetter: phaseGetter}
 	}
 
-	return nil
+	if len(caps) == 0 {
+		return base
+	}
+
+	return &decorateTestCapable{Charger: base, caps: caps}
+}
+
+type decorateTestCapable struct {
+	api.Charger
+	caps map[reflect.Type]any
+}
+
+func (d *decorateTestCapable) Capability(typ reflect.Type) (any, bool) {
+	c, ok := d.caps[typ]
+	return c, ok
 }
 
 type decorateTestMeterEnergyImpl struct {
diff --git a/cmd/decorate/header.tpl b/cmd/decorate/header.tpl
index b3ea644a5..c242ec257 100644
--- a/cmd/decorate/header.tpl
+++ b/cmd/decorate/header.tpl
@@ -3,5 +3,7 @@ package {{.Package}}
 // Code generated by github.com/evcc-io/evcc/cmd/tools/decorate.go. DO NOT EDIT.
 
 import (
+	"reflect"
+
 	"github.com/evcc-io/evcc/api"
 )
diff --git a/go.mod b/go.mod
index 2d5e09d25..deb8e7c68 100644
--- a/go.mod
+++ b/go.mod
@@ -74,7 +74,6 @@ require (
 	github.com/mitchellh/go-homedir v1.1.0
 	github.com/modelcontextprotocol/go-sdk v1.4.1
 	github.com/muka/go-bluetooth v0.0.0-20240701044517-04c4f09c514e
-	github.com/mxschmitt/golang-combinations v1.2.0
 	github.com/nicholas-fedor/shoutrrr v0.14.0
 	github.com/nicksnyder/go-i18n/v2 v2.6.1
 	github.com/olekukonko/tablewriter v1.1.4
diff --git a/go.sum b/go.sum
index c6d2c2f4b..72336242b 100644
--- a/go.sum
+++ b/go.sum
@@ -535,8 +535,6 @@ github.com/munnerz/goautoneg v0.0.0-20191010083416-a7dc8b61c822/go.mod h1:+n7T8m
 github.com/mwitkow/go-conntrack v0.0.0-20161129095857-cc309e4a2223/go.mod h1:qRWi+5nqEBWmkhHvq77mSJWrCKwh8bxhgT7d/eI7P4U=
 github.com/mwitkow/go-conntrack v0.0.0-20190716064945-2f068394615f h1:KUppIJq7/+SVif2QVs3tOP0zanoHgBEVAwHxUSIzRqU=
 github.com/mwitkow/go-conntrack v0.0.0-20190716064945-2f068394615f/go.mod h1:qRWi+5nqEBWmkhHvq77mSJWrCKwh8bxhgT7d/eI7P4U=
-github.com/mxschmitt/golang-combinations v1.2.0 h1:V5E7MncIK8Yr1SL/SpdqMuSquFsfoIs5auI7Y3n8z14=
-github.com/mxschmitt/golang-combinations v1.2.0/go.mod h1:RCm5eR03B+JrBOMRDLsKZWShluXdrHu+qwhPEJ0miBM=
 github.com/nats-io/jwt v0.3.0/go.mod h1:fRYCDE99xlTsqUzISS1Bi75UBJ6ljOJQOAAu5VglpSg=
 github.com/nats-io/jwt v0.3.2/go.mod h1:/euKqTS1ZD+zzjYrY7pseZrTtWQSjujC7xjPc8wL6eU=
 github.com/nats-io/nats-server/v2 v2.1.2/go.mod h1:Afk+wRZqkMQs/p45uXdrVLuab3gwv3Z8C4HTBu8GD/k=
